In [1]:
import pandas as pd
import numpy as np
import os 

In [2]:
path = r'/Users/danielmccormick/Desktop/CareerFoundry/Advanced Analytics & Dashboard Design/6.1_ElectricVehicles/Data'

In Excel, the data has already been transformed from five tables, individual tables for cars, vans, buses, and trucks with sales, stock, % of sales, and % of stock data, and one table with data on charging stations, into three tables. One with sales and stock information (df_sales_stock/df_sales_stock_imputed) one with % of sales and % of stock information (df_shares/df_shares_imputed), and one with information on charging stations (df_chargers/df_chargers_imputed).

During the process of transforming the data in Excel, if there were any missing data points for the period of 2010-2022 (the period of the data set), a value of zero was imputed. Those tables were then saved for further analysis. They are the dataframes with "imputed" at the end of the dataframe name. In order to have data sets that only include data explicitly reported by the IEA (without the imputed values), there are also data sets where the imputed values were then filtered back out. These are the dataframes without "imputed" in the name. This script will perform data cleaning and basic exploratory analysis on both data sets. 

In [3]:
df_sales_stock = pd.read_csv(os.path.join(path, 'EV_Sales_Stock.csv'), index_col = False)

In [4]:
df_sales_stock.head()

region vehicle_type powertrain_type  year  ev_sales  ev_stock
0  Australia         Cars             BEV  2011        49        49
1  Australia         Cars             BEV  2012       170       220
2  Australia         Cars            PHEV  2012        80        80
3  Australia         Cars             BEV  2013       190       410
4  Australia         Cars            PHEV  2013       100       180

In [5]:
df_shares = pd.read_csv(os.path.join(path, 'EV_Shares.csv'), index_col = False)

In [6]:
df_shares.head()

region vehicle_type  year  ev_sales_share  ev_stock_share
0  Australia         Cars  2011         0.00007         0.00000
1  Australia         Cars  2012         0.00030         0.00003
2  Australia         Cars  2013         0.00034         0.00005
3  Australia         Cars  2014         0.00160         0.00017
4  Australia         Cars  2015         0.00200         0.00032

In [7]:
df_chargers = pd.read_csv(os.path.join(path, 'EV_Charging_Stations.csv'), index_col = False)

In [8]:
df_chargers.head()

region             charger_type  year  charger_count
0  Australia  Publicly Available Slow  2017          440.0
1  Australia  Publicly Available Fast  2017           40.0
2  Australia  Publicly Available Slow  2018          670.0
3  Australia  Publicly Available Fast  2018           61.0
4  Australia  Publicly Available Slow  2019         1700.0

In [9]:
df_sales_stock_imputed = pd.read_csv(os.path.join(path, 'EV_Sales_Stock_imputed.csv'), index_col = False)

In [10]:
df_sales_stock_imputed.head()

region vehicle_type powertrain_type  year  ev_sales  ev_stock
0  Australia         Cars             BEV  2010         0         0
1  Australia         Cars            PHEV  2010         0         0
2  Australia         Cars             BEV  2011        49        49
3  Australia         Cars            PHEV  2011         0         0
4  Australia         Cars             BEV  2012       170       220

In [11]:
df_shares_imputed = pd.read_csv(os.path.join(path, 'EV_Shares_imputed.csv'), index_col = False)

In [12]:
df_shares_imputed.head()

region vehicle_type  year  ev_sales_share  ev_stock_share
0  Australia         Cars  2010         0.00000         0.00000
1  Australia         Cars  2011         0.00007         0.00000
2  Australia         Cars  2012         0.00030         0.00003
3  Australia         Cars  2013         0.00034         0.00005
4  Australia         Cars  2014         0.00160         0.00017

In [13]:
df_chargers_imputed = pd.read_csv(os.path.join(path, 'EV_Charging_Stations_imputed.csv'), index_col = False)

In [14]:
df_chargers_imputed.head()

region             charger_type  year  charger_count
0  Australia  Publicly Available Slow  2010            0.0
1  Australia  Publicly Available Fast  2010            0.0
2  Australia  Publicly Available Slow  2011            0.0
3  Australia  Publicly Available Fast  2011            0.0
4  Australia  Publicly Available Slow  2012            0.0

First step for cleaning the data is to confirm there is no missing or null values in the data set.

While examining and reformatting data in Excel it was clear there were no null values in the data set. Based on what was found during reformatting the data, rather than have a row with no reported data for a certain data point, the IEA just did not report anything for that time period. As explained above, in Excel a value of zero was imputed for any data point that was not reported for 2010-2022 (in the data sets with "imputed" in the dataframe name). Just to confirm that this was successful in Excel a check for null values will be run for all columns in the dataset.

In [15]:
df_sales_stock_imputed['region'].isnull().sum()

0

In [16]:
df_sales_stock['region'].isnull().sum()

0

In [17]:
df_sales_stock_imputed['vehicle_type'].isnull().sum()

0

In [18]:
df_sales_stock['vehicle_type'].isnull().sum()

0

In [19]:
df_sales_stock_imputed['powertrain_type'].isnull().sum()

0

In [20]:
df_sales_stock['powertrain_type'].isnull().sum()

0

In [21]:
df_sales_stock_imputed['year'].isnull().sum()

0

In [22]:
df_sales_stock['year'].isnull().sum()

0

In [23]:
df_sales_stock_imputed['ev_sales'].isnull().sum()

0

In [24]:
df_sales_stock['ev_sales'].isnull().sum()

0

In [25]:
df_sales_stock_imputed['ev_stock'].isnull().sum()

0

In [26]:
df_sales_stock['ev_stock'].isnull().sum()

0

In [27]:
df_shares_imputed['region'].isnull().sum()

0

In [28]:
df_shares['region'].isnull().sum()

0

In [29]:
df_shares_imputed['vehicle_type'].isnull().sum()

0

In [30]:
df_shares['vehicle_type'].isnull().sum()

0

In [31]:
df_shares_imputed['year'].isnull().sum()

0

In [32]:
df_shares['year'].isnull().sum()

0

In [33]:
df_shares_imputed['ev_sales_share'].isnull().sum()

0

In [34]:
df_shares['ev_sales_share'].isnull().sum()

0

In [35]:
df_shares_imputed['ev_stock_share'].isnull().sum()

0

In [36]:
df_shares['ev_stock_share'].isnull().sum()

0

In [37]:
df_chargers_imputed['region'].isnull().sum()

0

In [38]:
df_chargers['region'].isnull().sum()

0

In [39]:
df_chargers_imputed['charger_type'].isnull().sum()

0

In [40]:
df_chargers['charger_type'].isnull().sum()

0

In [41]:
df_chargers_imputed['year'].isnull().sum()

0

In [42]:
df_chargers['year'].isnull().sum()

0

In [43]:
df_chargers_imputed['charger_count'].isnull().sum()

0

In [44]:
df_chargers['charger_count'].isnull().sum()

0

Second step to confirm data is ready for analysis will be to perform a value_count on each variable and confirm the results align with expectations.

In [45]:
df_sales_stock_imputed['region'].value_counts()

Australia            130
Austria              130
Mexico               130
Netherlands          130
New Zealand          130
Norway               130
Other Europe         130
Poland               130
Portugal             130
Rest of the world    130
South Africa         130
Spain                130
Sweden               130
Switzerland          130
Turkiye              130
United Kingdom       130
USA                  130
Korea                130
Japan                130
Italy                130
EU27                 130
Belgium              130
Brazil               130
Canada               130
Chile                130
China                130
Denmark              130
Europe               130
Israel               130
Finland              130
France               130
Germany              130
Greece               130
Iceland              130
India                130
World                130
Name: region, dtype: int64

In [46]:
df_sales_stock['region'].value_counts()

World                123
EU27                 119
Europe               119
China                118
France               114
Belgium              107
United Kingdom       107
Denmark              105
Netherlands           98
Other Europe          95
Sweden                94
Germany               92
Norway                90
Poland                89
Austria               89
Spain                 87
Iceland               87
Canada                85
Italy                 83
Switzerland           81
Japan                 79
New Zealand           76
Finland               75
Portugal              71
Rest of the world     70
Korea                 66
Brazil                65
Israel                63
USA                   50
India                 49
Greece                49
Australia             46
Mexico                44
Chile                 42
Turkiye               42
South Africa          38
Name: region, dtype: int64

Value count on region for imputed data for 2010-2022, the count for each region is even. 13 years x 2 powertrain types x 5 vehicle types (cars, buses, trucks, vans, total) = 130. 
For the non-imputed data set it makes sense that there are no counts over 130 (would be impossible) and that the World (total) reports the most data points since it is an aggregation of all other data points. 
These results do not indicate any issues with this column in either data set. 

In [47]:
df_sales_stock_imputed['vehicle_type'].value_counts()

Cars      936
Buses     936
Vans      936
Trucks    936
Total     936
Name: vehicle_type, dtype: int64

In [48]:
df_sales_stock['vehicle_type'].value_counts()

Total     856
Cars      851
Vans      487
Buses     438
Trucks    275
Name: vehicle_type, dtype: int64

The imputed vehicle type data has an even amount of each type. 36 countries x 2 powertrain types X 13 years = 936. For the non-imputed vehicle type data the results again make sense as there is no value higher than 936, and the total summation for all vehicles has the most values. 

In [49]:
df_sales_stock_imputed['powertrain_type'].value_counts()

BEV     2340
PHEV    2340
Name: powertrain_type, dtype: int64

In [50]:
df_sales_stock['powertrain_type'].value_counts()

BEV     1768
PHEV    1139
Name: powertrain_type, dtype: int64

Imputed powertrain type data has even number which make sense, 36 regions X 5 vehicle types X 13 years = 2340. The non-imputed powertain type data also makes sense as there are no values above 2340. There is much more reported data on battery electric vehicles than plug-in hybrids.

In [51]:
df_sales_stock_imputed['year'].value_counts()

2010    360
2011    360
2012    360
2013    360
2014    360
2015    360
2016    360
2017    360
2018    360
2019    360
2020    360
2021    360
2022    360
Name: year, dtype: int64

In [52]:
df_sales_stock['year'].value_counts()

2022    284
2021    281
2020    277
2019    267
2018    252
2017    247
2016    233
2015    218
2014    199
2013    193
2012    182
2011    155
2010    119
Name: year, dtype: int64

Imputed year data in the sales/stock table even again, 36 regions x 5 vehicle types X 2 powertrain types = 360. Non-imputed year data also makes sense as there are no values over 360, and the values increase each year. This makes logical sense as more and more sales and stock data was reported each year of the data set.

In [53]:
df_shares_imputed['region'].value_counts()

Australia            52
Austria              52
Mexico               52
Netherlands          52
New Zealand          52
Norway               52
Other Europe         52
Poland               52
Portugal             52
Rest of the world    52
South Africa         52
Spain                52
Sweden               52
Switzerland          52
Turkiye              52
United Kingdom       52
USA                  52
Korea                52
Japan                52
Italy                52
EU27                 52
Belgium              52
Brazil               52
Canada               52
Chile                52
China                52
Denmark              52
Europe               52
Israel               52
Finland              52
France               52
Germany              52
Greece               52
Iceland              52
India                52
World                52
Name: region, dtype: int64

In [54]:
df_shares['region'].value_counts()

World                52
Belgium              52
United Kingdom       52
China                52
EU27                 52
Europe               52
France               52
Netherlands          52
Austria              52
Denmark              51
Iceland              49
Other Europe         48
Italy                42
Sweden               42
Poland               42
Japan                42
Spain                40
Canada               40
Norway               39
Germany              38
Korea                37
Switzerland          35
New Zealand          34
Brazil               32
Portugal             31
Rest of the world    31
Finland              30
India                28
Israel               26
Greece               20
Mexico               16
USA                  13
Chile                12
Australia            12
Turkiye              11
South Africa         10
Name: region, dtype: int64

Imputed region data in the shares table is all even, 4 vehicle types x 13 years = 52 (no powertrain type or "Total" vehicle type in Shares data). Non-imputed region data makes sense as there is no value over 52 and World and EU27 both have the max amount which makes sense as World is a total of all counties, and EU27 is a total of European countries, some of which have the max amount of values at 52. 

In [55]:
df_shares_imputed['vehicle_type'].value_counts()

Cars      468
Buses     468
Vans      468
Trucks    468
Name: vehicle_type, dtype: int64

In [56]:
df_shares['vehicle_type'].value_counts()

Cars      451
Vans      323
Buses     308
Trucks    237
Name: vehicle_type, dtype: int64

Imputed vehicle type data from shares table makes sense as we have even count for each, and 36 regions x 13 years = 468. Non-imputed vehicle type data makes sense as there is no value above 468. Cars have the most reported data points in the non-imputed data. 

In [57]:
df_shares_imputed['year'].value_counts()

2010    144
2011    144
2012    144
2013    144
2014    144
2015    144
2016    144
2017    144
2018    144
2019    144
2020    144
2021    144
2022    144
Name: year, dtype: int64

In [58]:
df_shares['year'].value_counts()

2022    124
2020    123
2021    123
2019    119
2018    114
2017    109
2016    103
2015     95
2014     89
2013     88
2012     85
2011     80
2010     67
Name: year, dtype: int64

Imputed data for year in the shares data set makes sense. The values are all even, with 36 regions X 4 vehicle types = 144. Non-imputed year data also makes sense as there is no value above 144 and the amount of reported data increased in each year in order from 2010 to 2022 (both 2020 and 2021 had 123 reported values). 

In [59]:
df_chargers_imputed['region'].value_counts()

Australia         26
South Africa      26
Mexico            26
Netherlands       26
New Zealand       26
Norway            26
Poland            26
Portugal          26
Spain             26
Austria           26
Sweden            26
Switzerland       26
Thailand          26
Turkiye           26
United Kingdom    26
USA               26
Korea             26
Japan             26
Italy             26
Israel            26
Indonesia         26
India             26
Iceland           26
Greece            26
Germany           26
France            26
Finland           26
Denmark           26
China             26
Chile             26
Canada            26
Brazil            26
Belgium           26
Total             26
Name: region, dtype: int64

In [60]:
df_chargers['region'].value_counts()

Total             26
Norway            25
Portugal          24
Chile             24
Netherlands       24
Austria           24
United Kingdom    23
Germany           22
Switzerland       22
Italy             22
Japan             22
Denmark           22
France            22
Canada            22
USA               21
Spain             21
Belgium           20
Finland           20
Greece            19
Poland            18
China             18
Iceland           17
Turkiye           17
Sweden            16
Mexico            15
Korea             14
Israel            14
India             13
South Africa      12
Thailand          12
Australia         12
New Zealand       11
Brazil            10
Indonesia          6
Name: region, dtype: int64

Region for the chargers data makes sense, as all values are even and 13 years x 2 charger types = 26. Non-imputed region data makes sense as there are no values above 26 and the region with the most reported values is the total which is a summation of all the individual regions. 

In [61]:
df_chargers_imputed['charger_type'].value_counts()

Publicly Available Slow    442
Publicly Available Fast    442
Name: charger_type, dtype: int64

In [62]:
df_chargers['charger_type'].value_counts()

Publicly Available Slow    317
Publicly Available Fast    313
Name: charger_type, dtype: int64

Imputed data for charger types makes sense, 34 regions X 13 years = 442. Non-imputed charger type data makes sense as there is no value about 442. There is slightly more reported data on slow chargings stations than fast charging stations. 

In [63]:
df_chargers_imputed['year'].value_counts()

2010    68
2011    68
2012    68
2013    68
2014    68
2015    68
2016    68
2017    68
2018    68
2019    68
2020    68
2021    68
2022    68
Name: year, dtype: int64

In [64]:
df_chargers['year'].value_counts()

2022    68
2020    67
2021    67
2018    66
2019    66
2017    63
2016    54
2015    49
2014    44
2013    37
2012    29
2011    14
2010     6
Name: year, dtype: int64

Imputed data for year in the charging stations data makes sense, 34 regions x 2 charger types = 68. Non-imputed year data makes sense as there are no values above 68 and the amount of reported values increased each year from 2010-2022 (2018-2019 and 2020-2021 both had the same amount of data points reported). 

Based on the value counts performed on all of the categorical variables there are no issues with the data sets so far that would cause issues in the electric vehicle analysis

The next step will be to peform basic statistical exploratory analysis on each quantitative variable to confirm the results make logical sense based on what the data point represents. The script used returns the normal statistics from describe() but converts them to not be in scientific notation. 

In [65]:
df_sales_stock_imputed.describe().apply(lambda x: x.apply('{0:.5f}'.format))

year       ev_sales        ev_stock
count  4680.00000     4680.00000      4680.00000
mean   2016.00000    30910.24274     85711.73803
std       3.74206   253177.36359    666840.70849
min    2010.00000        0.00000         0.00000
25%    2013.00000        0.00000         0.00000
50%    2016.00000       12.00000        50.00000
75%    2019.00000     1500.00000      4400.00000
max    2022.00000  7721000.00000  19930000.00000

- Count of 4680 for all makes sense.
- For year the mean and 50% percentile of 2016 make sense since all missing data points were imputed in this data set. 
- For year the min of 2010 and max of 2022 make sense since the time frame in 2010-2022
- For year the 25% and 75% percentiles being 2013 and 2019 make sense
- For year standard deviation of about 3.7 seems reasonable 
- For ev_sales and ev_Stock the means seem reasonable. These two measurements are being skewed lower by the significant amount of values that were imputed with zero in this data set. 
- For ev_sales and ev_stock standard deviation seems reasonable based on the min and max and number of values for each data point. 
- For ev_sales and ev_stock min and 25% percentile of 0 make sense since there were a significant amount of data points imputed with zero in the data set. 
- For ev_sales and ev_stock the median and 75% percentile seem low but reasonable. These too are being skewed by the significant amount of imputed zeros. 
- For ev_sales and ev_stock the maximum values seems reasonable at about 7.7M for sales and 19.9M for stock. These are probably the total sales and stock for the entire world in the most recent year of the data set. It also makes sense that the sales are well below the stock as it would be impossible to have more sales than available vehicles. 

In [66]:
df_sales_stock.describe().apply(lambda x: x.apply('{0:.5f}'.format))

year       ev_sales        ev_stock
count  2907.00000     2907.00000      2907.00000
mean   2016.81080    49762.61988    137987.93739
std       3.58026   319793.78460    841881.98327
min    2010.00000        0.00000         0.00000
25%    2014.00000       27.50000       100.00000
50%    2017.00000      580.00000      1500.00000
75%    2020.00000     8705.50000     21099.50000
max    2022.00000  7721000.00000  19930000.00000

- This data set contains the same information as the previous, except it excludes all the rows where both the ev_sales and ev_stock values were imputed with zero. 
- Makes sense that all three values have the same count and that it is less than the previous data set as it excludes fully imputed rows
- For year it makes sense that the mean is not a whole number and is higher than in the previous data set. There is more data reported in the later years of the data set which results in the mean for year being higher in this data set. 
- Year standard deviation is slightly lower but is close to the previous data set. This also makes sense since the lack of imputed values would decrease the spread of the data. 
- Year min and max are unchanged, while the 25% and 75% percentile are both higher which again makes sense since there is more reported data in the later years of the data set when imputed values are not included. 
- For ev_sales and ev_stock the mean seems reasonable for both and it makes sense that it is larger than in the previous data set since there are not nearly as many values of zero in this data. 
- For ev_sales and ev_stock it makes sense that the standard deviation is larger than in the previous data set because the lack of many imputed values of zero would increase the average distance of the data points from the mean. 
- For ev_sales and ev_stock min and max not changing from the previous data set makes sense. The reason the minimum is zero even though the imputed values were not included in this data is that if for one row of data the ev_sales was zero but the ev_stock was not, then that row remained in the data set. The opposite it true as well, if ev_stock was zero and ev_sales was not then it was also incuded in this data set. (Note: it logically does not make sense for there to be a row with an ev_stock of zero but ev_sales greater than zero. Further analysis of this question will be done below). 
- 25% and 75% percentiles both seem reasonable and it makes sense that they are higher than the previous data set since there are not as many values of zero in this data set. 

In [67]:
df_shares_imputed.describe().apply(lambda x: x.apply('{0:.5f}'.format))

year ev_sales_share ev_stock_share
count  1872.00000     1872.00000     1872.00000
mean   2016.00000        0.02126        0.00413
std       3.74266        0.07224        0.01539
min    2010.00000        0.00000        0.00000
25%    2013.00000        0.00000        0.00000
50%    2016.00000        0.00052        0.00014
75%    2019.00000        0.00910        0.00190
max    2022.00000        0.88000        0.27000

- For year all of the results make sense for the same reasons listed above for the other imputed data set. 
- For ev_sales_share and ev_stock_share the very low mean, minimum and 25% percentile of 0, and very low median and 75% percentiles all make sense as the sales share and stock share of electric vehicles has historically been low, and these values are also being skewed even lower by a significant amount of imputed values of zero in this data set. 
- The stadard deviations of both ev_sales_share and ev_stock_share seem reasonable. 
- The maximum value of 0.88 for ev_sales_share means that for one combination of region, vehicle type, and year, 88% of all vehicles sold were electric. (Note: this seems high and will be analyzed further below). 
- The maximum value of 0.27, or 27% for ev_stock_share seems high but more resonable than the 88% noted above. 

In [69]:
df_shares.describe().apply(lambda x: x.apply('{0:.5f}'.format))

year ev_sales_share ev_stock_share
count  1319.00000     1319.00000     1319.00000
mean   2016.65656        0.03017        0.00586
std       3.65315        0.08449        0.01806
min    2010.00000        0.00000        0.00000
25%    2014.00000        0.00032        0.00009
50%    2017.00000        0.00340        0.00071
75%    2020.00000        0.01900        0.00420
max    2022.00000        0.88000        0.27000

- Makes sense that the count is lower than the previous data set
- For year makes sense min and max are the same as the previous data set. 
- For year makes sense that mean, median, and 25% and 75% percentiles are all larger than the previous data set since there are not as many values of zero in this data set. 
- For year makes sense that the standard deviation is lower (very slightly) than the previous data set since the lack of imputed values reduces the spread of the data set. 
- For ev_sales_share and ev_stock_share it makese sense that min and max are the same as the previous data set.
- For ev_sales_share and ev_stock_share it makes sense that standard deviation is larger than in the previous data set because the lack of many imputed values of zero would increase the average distance of the data points from the mean. 
- For ev_sales_share and ev_stock_share the mean, median, and 25% and 75% percentiles all seem reasonable. It makes sense that every one of these values is higher for this data set than the previous one because of the lack of imputed values of zero. 

In [70]:
df_chargers_imputed.describe().apply(lambda x: x.apply('{0:.5f}'.format))

year  charger_count
count   884.00000      884.00000
mean   2016.00000    18611.38846
std       3.74378   107353.96068
min    2010.00000        0.00000
25%    2013.00000        0.00000
50%    2016.00000      160.00000
75%    2019.00000     2600.00000
max    2022.00000  1782810.00000

- Results for year are reasonable for the same reasons listed above for the other two imputed data sets. 
- For charger_count makes sense the count is the same as the count for year 
- The standard deviation of about 107k seems a bit high but is probably just due to the large variation of values in this data set. 
- The mean charger count seems reasonable. It is being skewed by the imputed values of zero. 
- The min and 25% percentile being zero makes sense due to the significant amount of imputed valued in this data set. 
- The median and 75% percentiles both seem reasonable. They too are being skewed lower by the imputed values. 
- The max value of 1.8M seems reasonable. it is most likely the amount for the total of all regions in this data set for the most recent year of the data set.

In [71]:
df_chargers.describe().apply(lambda x: x.apply('{0:.5f}'.format))

year  charger_count
count   630.00000      630.00000
mean   2017.48254    26115.02762
std       3.16964   126422.01873
min    2010.00000        0.10000
25%    2015.00000      110.00000
50%    2018.00000     1000.00000
75%    2020.00000     5950.00000
max    2022.00000  1782810.00000

- Makes sense that the count is lower than the previous data set
- For year makes sense min and max are the same as the previous data set. 
- For year makes sense that mean, median, and 25% and 75% percentiles are all larger than the previous data set since there are not as many values of zero in this data set. 
- For year makes sense that the standard deviation is lower than the previous data set since the lack of imputed values reduces the spread of the data set. 
- For charger_count it makese sense that min and max are the same as the previous data set.
- For charger_count it makes sense that standard deviation is larger than in the previous data set because the lack of many imputed values of zero would increase the average distance of the data points from the mean. This value does also seem high but not unreasonbly so.
- For charger_count the mean, median, and 25% and 75% percentiles all seem reasonable. It makes sense that every one of these values is higher for this data set than the previous one because of the lack of imputed values of zero. 

As noted above, it seems illogical that there would be any rows in the sales and stock table where the stock value would be zero, but there would be vehicle sales, since it seems impossible for there to be sales of vehicles when there is no stock available. 

The below query sets the max number of rows shown to 85 so the full result is shown, and then prints all rows from the sales and stock table for specific vehicle types (not total) where there is zero stock but more than zero sales. 

In [98]:
pd.set_option('display.max_rows', 85)
df_sales_stock[(df_sales_stock['ev_stock']== 0) & (df_sales_stock['ev_sales']> 0) & (df_sales_stock['vehicle_type'].isin(['Cars', 'Buses', 'Vans', 'Trucks']))]

region vehicle_type powertrain_type  year  ev_sales  ev_stock
73          Brazil         Cars             BEV  2012        74         0
74          Brazil         Cars            PHEV  2012        18         0
75          Brazil         Cars             BEV  2013       130         0
76          Brazil         Cars            PHEV  2013        33         0
590         Poland         Cars             BEV  2010        15         0
591         Poland         Cars            PHEV  2010         1         0
592         Poland         Cars             BEV  2011        43         0
593         Poland         Cars             BEV  2012        44         0
594         Poland         Cars            PHEV  2012         4         0
595         Poland         Cars             BEV  2013        31         0
596         Poland         Cars            PHEV  2013         6         0
597         Poland         Cars             BEV  2014       140         0
598         Poland         Cars            PHEV  2014        61         0
600         Poland         Cars            PHEV  2015       150         0
602         Poland         Cars            PHEV  2016       400         0
604         Poland         Cars            PHEV  2017       640         0
606         Poland         Cars            PHEV  2018       760         0
753        Turkiye         Cars             BEV  2012        92         0
755        Turkiye         Cars            PHEV  2013         1         0
757        Turkiye         Cars            PHEV  2014        22         0
759        Turkiye         Cars            PHEV  2015       110         0
886         Brazil        Buses             BEV  2010         3         0
887         Brazil        Buses             BEV  2011         3         0
888         Brazil        Buses             BEV  2012        94         0
889         Brazil        Buses             BEV  2013       120         0
890         Brazil        Buses             BEV  2015        13         0
891         Brazil        Buses             BEV  2016        15         0
892         Brazil        Buses             BEV  2017         2         0
895         Brazil        Buses             BEV  2020        55         0
897         Brazil        Buses             BEV  2022        37         0
898         Canada        Buses             BEV  2016         5         0
899         Canada        Buses             BEV  2017        11         0
1162        Poland        Buses             BEV  2018        63         0
1172      Portugal        Buses            PHEV  2014         2         0
1174      Portugal        Buses            PHEV  2017         2         0
1195         Spain        Buses             BEV  2014        11         0
1196         Spain        Buses             BEV  2015         2         0
1197         Spain        Buses             BEV  2016        27         0
1504        Greece         Vans            PHEV  2020         3         0
1534        Israel         Vans            PHEV  2021         2         0
1536        Israel         Vans            PHEV  2022        18         0
1540         Italy         Vans            PHEV  2012         3         0
1542         Italy         Vans            PHEV  2013         1         0
1547         Italy         Vans            PHEV  2017         2         0
1603        Norway         Vans             BEV  2010         1         0
1632  Other Europe         Vans            PHEV  2019         3         0
1639        Poland         Vans             BEV  2010         3         0
1640        Poland         Vans             BEV  2011         3         0
1641        Poland         Vans             BEV  2012        16         0
1642        Poland         Vans             BEV  2013        10         0
1643        Poland         Vans             BEV  2014         8         0
1645        Poland         Vans            PHEV  2015         2         0
1657      Portugal         Vans             BEV  2011         5         0
1658      Portug

It is only 80 rows of the total data set where this occurs, and it seems to be mostly small values for sales for these rows. There is not anything specific listed at the data source on why this would be occurring. It could be the vehicles were purchased from an international source within the country so that the stock was technically not in the country before the sale occurred. Regardless these rows should not cause a significant issue for the analysis and will be left how they are. 

As noted above it was not impossible, but surprising, to see a ev_sales_share value of 88%. This meant that for a full year in one of the regions, 88% of vehicles of one vehicle_type sold were electric. It is worth checking on where this occurred to confirm if it is seems reasonable. 

In [73]:
df_shares[df_shares['ev_sales_share'] == 0.88]

region vehicle_type  year  ev_sales_share  ev_stock_share
301  Norway         Cars  2022            0.88            0.27

It is interesting but unsurprising to see that the the max value for ev_sales and ev_share occurred in the same row, for Norway in the most recent year of data. After some quick research it is clear that Norway is the global leader is eletric car adoption so these values are not surprsing or an indication that there is an issue with the data. 

Based on the above analysis, these six data sets are clean and prepared for further analysis. 